pkg/controller/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -0,0 +1,262 @@
+package nstemplateset
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/codeready-toolchain/toolchain-common/pkg/condition"
+
+	"github.com/codeready-toolchain/member-operator/pkg/apis"
+	"github.com/codeready-toolchain/toolchain-common/pkg/test"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+const (
+	namespaceName = "toolchain-member"
+)
+
+func TestFindNamespace(t *testing.T) {
+	namespaces := []corev1.Namespace{
+		{ObjectMeta: metav1.ObjectMeta{Name: "johnsmith-dev"}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "johnsmith-code"}},
+	}
+
+	t.Run("found", func(t *testing.T) {
+		nsName := "johnsmith-dev"
+		namespace, found := findNamespace(namespaces, nsName)
+		assert.True(t, found)
+		assert.Equal(t, nsName, namespace.GetName())
+	})
+
+	t.Run("not_found", func(t *testing.T) {
+		nsName := "johnsmith-stage"
+		_, found := findNamespace(namespaces, nsName)
+		assert.False(t, found)
+	})
+}
+
+func TestNextMissingNamespace(t *testing.T) {
+	username := "johnsmith"
+	userNamespaces := []corev1.Namespace{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "johnsmith-dev", Labels: map[string]string{"revision": "rev1"},
+			},
+			Status: corev1.NamespaceStatus{Phase: corev1.NamespaceActive},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "johnsmith-code",
+			},
+			Status: corev1.NamespaceStatus{Phase: corev1.NamespaceActive},
+		},
+	}
+	tcNamespaces := []toolchainv1alpha1.Namespace{
+		{Type: "dev", Revision: "rev1"},
+		{Type: "code", Revision: "rev1"},
+		{Type: "stage", Revision: "rev1"},
+	}
+
+	t.Run("mistach_revision", func(t *testing.T) {
+		// test
+		tcNS, userNS, found := nextMissingNamespace(tcNamespaces, userNamespaces, username)
+
+		assert.True(t, found)
+		assert.Equal(t, "code", tcNS.Type)
+		assert.Equal(t, "johnsmith-code", userNS.GetName())
+	})
+
+	t.Run("found_next_namespace", func(t *testing.T) {
+		userNamespaces[1].Labels = map[string]string{"revision": "rev1"}
+
+		// test
+		tcNS, userNS, found := nextMissingNamespace(tcNamespaces, userNamespaces, username)
+
+		assert.True(t, found)
+		assert.Equal(t, "stage", tcNS.Type)
+		assert.Nil(t, userNS)
+	})
+
+	t.Run("not_found_next_namespace", func(t *testing.T) {
```

Comment:
this subtest depends on the setup of the previous subtest

Reply:
removing subtest.